### PR TITLE
changes for fetch max user and groups in single call and also remove …

### PR DESCRIPTION
**Pull request by prakashcc**
## Type of change
Bug Fix- Performance Issue

Below are the fixes -
1.) Currently, API is not using its full potential and only fetching 100 users and groups which is the default. this is 9 times in comparison with current
2.) inside the getUser function it is making multiple calls to fetch the same user data which I have combined and now getting in a single call. This will make this function from log (2n) to log(n). 

## Migration notes

> Describe migration notes if any

## Extra info

> Add any additional information

### CODE REVIEW

File: sources/azureactivedirectory-source/src/azureactivedirectory.ts [hunk 1]
```diff
@@ -128,12 +118,13 @@ export class AzureActiveDirectory {
       } catch (error) {
         this.logger.error(error.toString());
       }
+      console.log(item);
```
Comment (tovbinm) [inline sources/azureactivedirectory-source/src/azureactivedirectory.ts:121]: remove the console log

File: sources/azureactivedirectory-source/src/azureactivedirectory.ts [hunk 2]
```diff
@@ -104,20 +104,10 @@ export class AzureActiveDirectory {
   }
 
   async *getUsers(): AsyncGenerator<User> {
-    const res = await this.httpClient.get<UserResponse>('users');
+    const res = await this.httpClient.get<UserResponse>(
+      'users/?$select=Department,postalCode,createdDateTime,identities,streetAddress&$top=999'
```
Comment (tovbinm) [inline sources/azureactivedirectory-source/src/azureactivedirectory.ts:108]: why aren't we paginating?

File: sources/azureactivedirectory-source/src/azureactivedirectory.ts (hunk 1 above)
Comment (prakashcc) [reply sources/azureactivedirectory-source/src/azureactivedirectory.ts]: i removed it

File: sources/azureactivedirectory-source/src/azureactivedirectory.ts (hunk 2 above)
Comment (prakashcc) [reply sources/azureactivedirectory-source/src/azureactivedirectory.ts:108]: I am going to add pagination next but these fixes will still be relevant after adding the pagination.

Comment (tovbinm) [reply sources/azureactivedirectory-source/src/azureactivedirectory.ts:108]: I would make the page size configurable with a sane default. Would 999 be a good one? or it's safer to go with a smaller value, eg. 250, 500?

Comment (prakashcc) [reply sources/azureactivedirectory-source/src/azureactivedirectory.ts:108]: this size is recommended by the azure-ad team and I have tested it with 999 users working fine. As soon as I will implement the pagination we will make it as custom properties like we have client_id etc.

Comment (tovbinm) [reply sources/azureactivedirectory-source/src/azureactivedirectory.ts:108]: @huongtn FYI

Comment (tovbinm) [reply sources/azureactivedirectory-source/src/azureactivedirectory.ts:108]: 1. I would like to have you add pagination 
2. I think there is a better way to pass query arguments than this (see params - https://masteringjs.io/tutorials/axios/get)

Comment (prakashcc) [reply sources/azureactivedirectory-source/src/azureactivedirectory.ts]: @tovbinm based on your suggestion. changes are completed.

**Review by tovbinm — APPROVED** [review]
just one comment left, otherwise lgtm!

**Comment by cjwooo** [comment]
Please also fix the failing test https://github.com/faros-ai/airbyte-connectors/runs/5994598081?check_suite_focus=true#step:6:65

File: sources/azureactivedirectory-source/src/azureactivedirectory.ts [hunk 3]
```diff
@@ -103,59 +103,90 @@ export class AzureActiveDirectory {
...
       }
-      yield item;
+      yield user;
     }
   }
 
   async *getGroups(): AsyncGenerator<Group> {
-    const res = await this.httpClient.get<GroupResponse>('groups');
-    for (const item of res.data.value) {
+    const maxResults = 999;
```
Comment (tovbinm) [inline sources/azureactivedirectory-source/src/azureactivedirectory.ts:163]: `async *getGroups(maxResults = 999)` same above

File: sources/azureactivedirectory-source/src/azureactivedirectory.ts [hunk 4]
```diff
@@ -103,59 +103,90 @@ export class AzureActiveDirectory {
...
-        item.streetAddress = extraUserInfo.data.streetAddress;
-        item.identities = extraUserInfo.data.identities;
+  private async *paginate<T>(path: string, param = {}): AsyncGenerator<T> {
+    let after = null;
+    do {
+      try {
+        const res = await this.httpClient.get<T[]>(path, param);
+        const linkHeader = res.data['@odata.nextLink'];
+        if (linkHeader) {
+          after = linkHeader.split('$skiptoken=').pop();
```
[the remark this answers is not in the document]
Comment (tovbinm) [reply sources/azureactivedirectory-source/src/azureactivedirectory.ts:113]: @prakashcc FYI ^

[the remark this answers is not in the document]
Comment (prakashcc) [reply sources/azureactivedirectory-source/src/azureactivedirectory.ts:113]: @tovbinm I tried to parse with the format-link-header but it is not working as per expectation. I am getting an empty object in return.

[the remark this answers is not in the document]
Comment (tovbinm) [reply sources/azureactivedirectory-source/src/azureactivedirectory.ts:113]: can you share an example of `linkHeader` value?

[the remark this answers is not in the document]
Comment (prakashcc) [reply sources/azureactivedirectory-source/src/azureactivedirectory.ts:113]: this is the "@odata.nextLink": "https://graph.microsoft.com/v1.0/users?$top=5&$skiptoken=X%274453707 ... 6633B900000000000000000000%27"

[the remark this answers is not in the document]
Comment (tovbinm) [reply sources/azureactivedirectory-source/src/azureactivedirectory.ts:113]: How about using the `URL`?
```
> new URL('https://graph.microsoft.com/v1.0/users?$top=5&$skiptoken=X%274453707 ... 6633B900000000000000000000%27')?.searchParams.get('$skiptoken')
"X'4453707 ... 6633B900000000000000000000'"
```

[the remark this answers is not in the document]
Comment (prakashcc) [reply sources/azureactivedirectory-source/src/azureactivedirectory.ts:113]: @tovbinm The suggested solution is working but as per my understanding, the split and pop are more memory optimized.

[the remark this answers is not in the document]
Comment (prakashcc) [reply sources/azureactivedirectory-source/src/azureactivedirectory.ts]: But I have made the changes based on your recommendation.

[the remark this answers is not in the document]
Comment (tovbinm) [reply sources/azureactivedirectory-source/src/azureactivedirectory.ts]: Memory is not an issue. I would like to have pagination work even when there an extra url argument / different order is returned. That's why URL parsing is preferred.

File: sources/azureactivedirectory-source/src/azureactivedirectory.ts [hunk 5]
```diff
@@ -1,5 +1,6 @@
 import axios, {AxiosInstance} from 'axios';
 import {AirbyteLogger, wrapApiError} from 'faros-airbyte-cdk/lib';
+import formatLinkHeader from 'format-link-header';
```
Comment (tovbinm) [inline sources/azureactivedirectory-source/src/azureactivedirectory.ts:3]: you dont need `formatLinkHeader`

Comment (prakashcc) [reply sources/azureactivedirectory-source/src/azureactivedirectory.ts]: @tovbinm fixed it sorry

**Review by tovbinm — APPROVED** [review]

**Comment by tovbinm** [comment]
test is failing, otherwise lgtm

**Pull request merged**